### ISSUE TITLE
Support single WebVTT subtitle format for HLS

### ISSUE THREAD

**Issue by Jason239** [google/shaka-packager#575]
As I know, HLS supports subtitles in single WebVTT format without any problems.
But now it isn't supported in Shaka packager. 
In addition, I hope it is possible to package the subtitle in single format(WebVTT) with segmented media(video and audio) in DASH and HLS.
For now if I try like that, then the error occurred like below.
`Inconsistent stream descriptor specification: segment_template should be specified for none or all stream descriptors.`

Otherwise if the segmented WebVTT format with MPD is supported, then it would be good.

**Comment by kqyang** [comment]
@Jason239 Thanks for filing this issue. It is an over-sight from our side. We thought it was not allowed to have a single WebVTT file in HLS. 

HLS specification actually allows SUBTITLES renditions to have a different target duration:

>    Each Media Playlist in each Variant Stream MUST have the same
      target duration.  The only exceptions are SUBTITLES Renditions and
      Media Playlists containing an EXT-X-I-FRAMES-ONLY tag, which MAY
      have different target durations if they have an EXT-X-PLAYLIST-
      TYPE of VOD.

We will work on supporting single WebVTT file in HLS. Added to milestone v2.5.

> Otherwise if the segmented WebVTT format with MPD is supported, then it would be good.

Unfortunately DASH allows side-loading only one file. See 

https://dashif.org/docs/DASH-IF-IOP-v4.1-clean.pdf

> 6.4.5. Guidelines for side-loaded TTML and WebVTT files
>
> Side-loaded TTML or WebVTT subtitles or caption files can be used by some players including dash.js. 
>
> Only one file for the full period is permitted, practically limiting this use case to non-live content.

**Comment by Jason239** [comment]
> Unfortunately DASH allows side-loading only one file.

I see, I didn't know it.
Okay, thanks !

**Comment by gtebbutt** [comment]
@kqyang Am I correct in understanding that there's currently no way to output shared DASH & HLS files if there's a text track included? 

Currently we're using the same single-file MP4 audio & video output for both, but it looks as though we'd need to use single-file DASH and segmented HLS if we add a text track. Presumably once milestone 2.5 is out then we can just add text to our existing pipeline and have it work with both formats?

**Comment by kqyang** [comment]
@gtebbutt Yes. Your understanding is correct.

You are welcome to contribute patch to Shaka Packager if you want it to happen sooner!

**Comment by bricemiramont** [comment]
Hi,
what I do is package for DASH with the subtitle tracks, then package for HLS but without subtitle tracks and then have a script generate a m3u playlist with 1 segment with the total length being the length of the media.
I add the subs playlists in the main playlist and append SUBTITLES="text" to the EXT-X-STREAM-INF line in the main playlist. "text" being the GROUP-ID of the subtitle tracks.

**Comment by kaharman** [comment]
@gtebbutt Is it works?

**Comment by bricemiramont** [comment]
Yes it does. The final (working) steps are a bit different to what I wrote earlier. It is in fact :
- generate pssh for widevine (with my content keys, contentID and signer)
- generate pssh for fairplay (with my content keys)
- package for audio and video tracks for fairplay and widevine in one command (with raw_key), cbcs
- use a bash command (perl -pi -e '...' movie.mpd) to replace </Period> with the Adaptation Set of my subtitles followed by </Period>. This is to add subtitles for dash
--- For HLS ---
- manually create playlist with my subtitle file and the total duration. exemple : 
```
#EXTM3U
#EXT-X-VERSION:6
#EXT-X-TARGETDURATION:122
#EXT-X-PLAYLIST-TYPE:VOD
#EXTINF:122.000,
movie3_subs_5_eng.vtt
#EXT-X-ENDLIST
```
- add a line in the main playlist, at the end of the file : 
`#EXT-X-MEDIA:TYPE=SUBTITLES,URI="movie3_subs_5_eng.m3u8",GROUP-ID="text",LANGUAGE="en",NAME="en",AUTOSELECT=YES`
- modify main playlist to add subtitle for each video track that I have
`#EXT-X-STREAM-INF:BANDWIDTH=3129651,AVERAGE-BANDWIDTH=2848283,CODECS="avc1.4d4028,mp4a.40.2",RESOLUTION=1920x1080,AUDIO="audio-group",SUBTITLES="text"`

this is executed in a loop for each subtitle track that I need. 
Being no bash guru, I build the bash script in nodejs (media metadata in json are a lot easier to manipulate in js for me).

**Comment by abdevnet** [comment]
@bricemiramont I'm building my subtitle file this way as well in a manual step after I run shaka packager.  However, I'm having an issue where my subtitles start to drift and get behind the actual dialog.  Did you experience this?  I'm curious what segment duration are you using for your video and audio files?  I'm using 3 seconds as the GOP in our original encodes is at 3 seconds.

**Comment by ScripterYoda** [comment]
Is there any plan for implementing this feature? It is quite a priority for our project.

**Comment by kqyang** [comment]
@ScripterYoda We do not have plans to implement it right now. Will you be interested in helping us implement it? We'll be happy to provide pointers.

**Comment by ScripterYoda** [comment]
Yes, I will try, so any pointers will be appreciated.

**Comment by kqyang** [comment]
@ScripterYoda It should mostly works now with just one more change:

The code requires `segment_template` to be specified for either all streams or none of the streams:
https://github.com/google/shaka-packager/blob/master/packager/packager.cc#L314. 

There can be an exception for WebVtt, e.g.

```c++
    if (on_demand_dash_profile != descriptor.segment_template.empty() &&
        // DASH allows text streams to be side-loaded.
        // HLS allows text streams to have a different target duration, including a single output segment.
        !(on_demand_dash_profile && GetOutputFormat(descriptor) == CONTAINER_WEBVTT)) {
      return Status(error::INVALID_ARGUMENT,
                    "Inconsistent stream descriptor specification: "
                    "segment_template should be specified for none or all "
                    "stream descriptors.");
    }
```

Can you try if it works with the change? If so, you are welcome to send a PR to us with test.
